cpp/cert/src/rules/CTR55-CPP/DoNotUseAnAdditiveOperatorOnAnIterator.ql

```diff
@@ -15,23 +15,44 @@ import cpp
 import codingstandards.cpp.cert
 import codingstandards.cpp.Iterators
 
-from ContainerIteratorAccess it
+/**
+ * any `.size()` check above our access
+ */
+predicate size_checked_above(ContainerIteratorAccess it, IteratorSource source) {
+  exists(STLContainer c, FunctionCall guardCall |
+    c.getACallToSize() = guardCall and
+    guardCall = it.getAPredecessor*() and
+    //make sure its the same container providing its size as giving the iterator
+    globalValueNumber(guardCall.getQualifier()) = globalValueNumber(source.getQualifier()) and
+    // and the size call we match must be after the assignment call
+    source.getASuccessor*() = guardCall
+  )
+}
+
+/**
+ * some loop check exists like: `iterator != end`
+ * where a relevant`.end()` call flowed into end
+ */
+predicate valid_end_bound_check(ContainerIteratorAccess it, IteratorSource source) {
+  exists(STLContainer c, Loop l, ContainerIteratorAccess otherAccess, IteratorSource end |
+    end = c.getAnIteratorEndFunctionCall() and
```

Comment:
Can we generalize this with `getANonConstIteratorEndFunctionCall` and `getAConstIteratorEndFunctionCall`?

Reply:
I dont understand the suggestion,
this predicate is already [returning the combination of those two you suggest](https://github.com/github/codeql-coding-standards/blob/main/cpp/common/src/codingstandards/cpp/Iterators.qll#L312)

Reply:
True, but it is limited to `STLContainer` iterators. I assume we want to include user defined iterators as well?

Reply:
valid, will extend `ContainerAccessWithoutRangeCheck` lib to include the more general end defn (similar to the more general size call)